### PR TITLE
 [common] Generalized MXFP8 fused kernels w.r.t. input tensor dimensions

### PR DESCRIPTION
**Pull request by Oleg-Goncharov**
# Description

Lifted restrictions on input tensor dimensions in MXFP8 fused quantize/dequantize kernels, allowing arbitrary number of rows (or the product of all tensor dimensions except the last of high-dimensional tensors).
The number of columns (or the dimensionality of the last dimension of high-dimensional tensors) must satisfy the memory alignment requirement for Tensor Memory Accelerator (TMA) to be a multiple of 16 bytes.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Removed the `isFullTile` requirement in the `mxfp8_quantize` and `mxfp8_dequantize` functions
- Added 16-byte alignment check when creating a TMA descriptor 
- MXFP8 Test Suite (quantize and dequantize): Added the alignment/padding requirement for dimensions of the tensors with scaling factors to be multiples of [128,4] for row-wise and [4,128] for column-wise scaling
- Modified `exp2f_rcp` function so it returns `1.0` if the `biased_exp==0` to avoid `NaN` in the `amax=0` scenario

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/cpp/operator/test_cast_mxfp8.cu [hunk 1]
```diff
@@ -397,10 +430,16 @@ void performTest_x2(const ProcessingMethod processing_method,
 }
 
 std::vector<std::pair<size_t, size_t>> matrix_sizes = {
+    {1, 16},
+    {3, 32},
+    {16, 48},
+    {21, 64},
+    {37, 96},
```
Comment (timmoon10) [inline tests/cpp/operator/test_cast_mxfp8.cu:437]: Do we need all these cases? It seems excessive to dedicate 60% of the tests to non-performant edge cases. I think it would be fine to go with `{1, 16}`, `{16, 48}`, `{993, 512}`.

File: transformer_engine/common/common.cu [hunk 2]
```diff
@@ -97,7 +97,15 @@ void create_2D_tensor_map(CUtensorMap &tensorMap, const SimpleTensor &tensor,
...
-  NVTE_CHECK(isPointerAligned(dataPtr, 16), "Tensor data must be 16B aligned");
+
+  constexpr int TMA_gmem_alignment = 16;  // Alignment of the global memory address
+  NVTE_CHECK(isPointerAligned(dataPtr, TMA_gmem_alignment),
+             "Tensor data pointer must be 16B aligned");
+
+  const int TMA_needed_size = TMA_gmem_alignment / type_size;
+  NVTE_CHECK(globalX % TMA_needed_size == 0, "Shape not supported. Expected multiple of " +
+                                                 std::to_string(TMA_needed_size) + ", got " +
+                                                 std::to_string(globalX));
```
Comment (timmoon10) [inline transformer_engine/common/common.cu:108]: `NVTE_CHECK` already converts to string internally. It can even print out `std::vector<T>`.
```suggestion
  const int TMA_needed_size = TMA_gmem_alignment / type_size;
  NVTE_CHECK(globalX % TMA_needed_size == 0, "Shape not supported. Expected multiple of " +
                                                 TMA_needed_size + ", got " +
                                                 globalX);
```

**Review by timmoon10 — APPROVED** [review]
Mostly LGTM. It looks like this is needed before we can merge https://github.com/NVIDIA/TransformerEngine/pull/1435.

File: tests/cpp/operator/test_cast_mxfp8.cu (hunk 1 above)
Comment (Oleg-Goncharov) [reply tests/cpp/operator/test_cast_mxfp8.cu:437]: Agree, we can remove them. I just tested different combinations of edge cases which overlap a bit

Comment (timmoon10) [reply tests/cpp/operator/test_cast_mxfp8.cu]: I think we went too far in the other direction. The existing test cases (`{128, 128}`, `{256, 256}`, `{768, 1024}`) were important to test since they're closer to the standard use-cases.

File: transformer_engine/common/common.cu (hunk 2 above)
Comment (timmoon10) [reply transformer_engine/common/common.cu:108]: Ah, it should be something like:
```suggestion
  const int TMA_needed_size = TMA_gmem_alignment / type_size;
  NVTE_CHECK(globalX % TMA_needed_size == 0, "Shape not supported. Expected multiple of ",
                                             TMA_needed_size, ", got ", globalX);
```
This is not high priority, but taking advantage of these convenience features makes it easier to write detailed error messages.

**Review by ptrendx — APPROVED** [review]

**Comment by timmoon10** [comment]
Pipeline 23264045

**Comment by timmoon10** [comment]
I see test failures when combined with https://github.com/NVIDIA/TransformerEngine/pull/1435: https://github.com/NVIDIA/TransformerEngine/pull/1435#issuecomment-2626453407
It seems to pass tests when run in isolation: pipeline 23274414

**Review by timmoon10 — APPROVED** [review]
LGTM. I still see errors when this is combined with #1435 (pipeline 23302072), but that seems related to MXFP8 gated activations, LayerNorm, RMSNorm. The MXFP8 quantize and dequantize kernels seem to be doing the right thing.
